service/storage_service.cc

```diff
@@ -5948,6 +5965,22 @@ future<raft_topology_cmd_result> storage_service::raft_topology_cmd_handler(shar
                     }
 
                     co_await ss._shared_token_metadata.stale_versions_in_use();
+
+                    auto&& tmeta = ss._shared_token_metadata.get()->tablets();
+                    for (auto&& [table_id, tmap] : tmeta.all_tables()) {
+                        for (auto&& [tid, trinfo] : tmap.transitions()) {
+                            if (trinfo.session_id) {
+                                if (trinfo.stage == locator::tablet_transition_stage::write_both_read_new) {
+                                    close_session(session_id(trinfo.session_id));
```

Comment:
I am trying to grasp how this is going to play out upon a coordinator change. On one hand session id is part of tablet metadata, so it's going to be loaded by the fresh coordinator, on the other, you don't wait for the new session id to be persisted in Raft before you close the previous session -  both events are happening in the background...

Could you formualte what you're trying to achieve in terms of invariants, so that one could try to validate them lookign at this code?

Reply:
Another concern is that going forward it would be rather difficult to ensure that these invariants are not broken:  in theory we could have some test coverage (a good idea anyway), but ideally, at least on the level of session tracking, we should assert them - e.g. that the session ids are monotonic, and that there are no more than two sessions against the same tablet (closed, established) at the same time.

Reply:
There are two pieces here.

One goal here is to make sure that local set of open sessions mirrors sessions stored in group0. It is group0 state which determines whether a given session is open or closed. If session_id is set in tablet metadata, it means we want this session to be open. Ideally, we should close all sessions which don't have entries in tablet metadata. Maybe that would be more robust against future changes. Here, I do something simpler but a bit more hacky, which is to close them in write_both_read_new, where we know we want the session to be closed but still keep it in tablet info so that this path can pick up the session_id for closing. We know every replica will observe this because the coordinator executes a barrier in write_both_read_new. An alternative approach would be clear the session immediately when entering write_both_read_new and here close all sessions which are not in group0 state. 

Another piece is that the token metadata barrier executed by the coordinator ensures that each replica sees the group0 state, so will also reflect intended state of sessions, and guarantees that closed sessions are drained.

The highest invariant (0) we want to preserve is that there are no side-effects from operations started in the streaming stage once the coordinator moved to the next stage. In this implementation, this is ensured by successful execution of the step `tablet_transition_stage::write_both_read_old`, which executes a barrier, which will close and drain sessions used for the streaming stage.

Reply:
During our calls you were also mentioning in-flight requests. Session ids are only checked on rpc entry, and in-flight requests may be stuck in the reactor. How is this approach addressing this problem?

Reply:
Another concern is that session map is independent on tablet map, at least in C++ code. The code assumes that a tablet can not disappear from system.tablets without its session being closed. But is this really the case? There is no strong reference in the code between the two, in theory we could just forget the UUID of the session forever if the tablet is deleted without the respective session being destroyed. This, of course, shouldn't happen, well, at least in theory..

Reply:
> During our calls you were also mentioning in-flight requests. Session ids are only checked on rpc entry, and in-flight requests may be stuck in the reactor. How is this approach addressing this problem?

OK, this is actually clear, since they will be holding the gate and you drain the gate in close_sessions(), we'll wait for them... There is an unfortunate similarity of this approach with locks however. We don't know what resource the write will be holding and in turn waiting on but yet we assume that draining will succeed. As we get too many such subsystems in place, we will end up waiting on draining a session which is waiting on draining a fence or through workload prioritization get into some priority inversion and wait for an a completely unrelated semaphore or a gate.

Reply:
> OK, this is actually clear, since they will be holding the gate and you drain the gate in close_sessions(), we'll wait for them... There is an unfortunate similarity of this approach with locks however. We don't know what resource the write will be holding and in turn waiting on but yet we assume that draining will succeed. As we get too many such fence/wait/drain subsystems in place, we will end up waiting on draining a session which is waiting on draining a fence or through workload prioritization get into some priority inversion and wait for an a completely unrelated semaphore or a gate.

Yes. What makes this problem less severe is the fact that this waiting only happens when there is a failed streaming, which is not a typical situation. When streaming is successful, there should be no guards alive for the closed session.

Reply:
> Another concern is that session map is independent on tablet map, at least in C++ code. The code assumes that a tablet can not disappear from system.tablets without its session being closed. But is this really the case? There is no strong reference in the code between the two, in theory we could just forget the UUID of the session forever if the tablet is deleted without the respective session being destroyed. This, of course, shouldn't happen, well, at least in theory..

Yes, this code needs to be changed to be more robust in closing sessions.